File: scripts/build-lpcc-ncmhce-form-b-v2-seeds.py
#!/usr/bin/env python3
"""Build LPCC NCMHCE Form B v2.0 admin-only answer key from parsed sources.

CLI only. Never prints correct letters to stdout in bulk.
Mirrors scripts/build-lpcc-ncmhce-form-a-v2-seeds.php for Form B.

Usage: python scripts/build-lpcc-ncmhce-form-b-v2-seeds.py
"""

from __future__ import annotations

import html
import json
import re
import sys
import xml.etree.ElementTree as ET
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
CAND_JSON = ROOT / "_tmp_lpcc_form_b_v2" / "parsed.json"
KEY_PARAS = ROOT / "_tmp_lpcc_form_b_v2_key" / "paras.txt"
KEY_XML = ROOT / "_tmp_lpcc_form_b_v2_key" / "unzip" / "word" / "document.xml"
REPORT = ROOT / "_tmp_lpcc_form_b_v2_key" / "match_report.txt"
ADMIN_PATH = ROOT / "includes" / "quiz-seeds" / "admin-only" / "lpcc-ncmhce-form-b-v2-answer-key.php"
NS = {"w": "http://schemas.openxmlformats.org/wordprocessingml/2006/main"}


def norm(text: str) -> str:
    text = html.unescape(text or "")
    text = text.replace("\u2018", "'").replace("\u2019", "'")
    text = text.replace("\u201c", '"').replace("\u201d", '"')
    text = text.replace("\u00a0", " ").replace("—", "-").replace("–", "-")
    return re.sub(r"\s+", " ", text).strip()


def php_export(value, level: int = 0) -> str:
    pad = "\t" * level
    pad1 = "\t" * (level + 1)
    if isinstance(value, list):
        if not value:
            return "array()"
        lines = ["array(\n"]
        for item in value:
            lines.append(f"{pad1}{php_export(item, level + 1)},\n")
        lines.append(f"{pad})")
        return "".join(lines)
    if isinstance(value, dict):
        if not value:
            return "array()"
        lines = ["array(\n"]
        for k, v in value.items():
            lines.append(f"{pad1}{php_export(str(k), level + 1)} => {php_export(v, level + 1)},\n")
        lines.append(f"{pad})")
        return "".join(lines)
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return str(value)
    escaped = str(value).replace("\\", "\\\\").replace("'", "\\'")
    return f"'{escaped}'"


def cell_text(cell) -> str:
    parts = []
    for t in cell.findall(".//w:t", NS):
        if t.text:
            parts.append(t.text)
    return norm("".join(parts))


def extract_paras() -> list[str]:
    tree = ET.parse(KEY_XML)
    root = tree.getroot()
    out = []
    for p in root.findall(".//w:body/w:p", NS):
        parts = []
        for t in p.findall(".//w:t", NS):
            if t.text:
                parts.append(t.text)
        out.append(norm("".join(parts)))
    KEY_PARAS.parent.mkdir(parents=True, exist_ok=True)
    KEY_PARAS.write_text("\n".join(out), encoding="utf-8")
    return out


def flatten_candidate(candidate: dict) -> list[dict]:
    items = []
    for case in candidate.get("cases", []):
        case_no = int(case.get("case_number") or 0)
        for sec in case.get("sections", []):
            sec_no = int(sec.get("section_number") or 0)
            stem = str(sec.get("stem") or "")
            first = True
            for q in sec.get("questions", []):
                num = int(q.get("number") or 0)
                items.append(
                    {
                        "case_number": case_no,
                        "case_title": str(case.get("title") or ""),
                        "section_number": sec_no,
                        "section_title": str(sec.get("title") or ""),
                        "section_stem": stem,
                        "is_section_first": first,
                        "number": num,
                        "question_text": str(q.get("question_text") or ""),
                        "options": dict(q.get("options") or {}),
                    }
                )
                first = False
    return items


def parse_table_meta() -> tuple[list[dict], list[str]]:
    tree = ET.parse(KEY_XML)
    root = tree.getroot()
    tables = root.findall(".//w:tbl", NS)
    table_meta: list[dict] = []
    control_bits: list[str] = []
    for ti, tbl in enumerate(tables, start=1):
        rows = []
        for tr in tbl.findall("./w:tr", NS):
            vals = [cell_text(tc) for tc in tr.findall("./w:tc", NS)]
            rows.append(" | ".join(vals))
        blob = " ".join(rows)
        if ti in (1, 146):
            control_bits.append(blob)
            continue
        if ti == 2:
            continue
        key = status = task = domain = ""
        m = re.search(r"Key:\s*([A-D])", blob, re.I)
        if m:
            key = m.group(1).lower()
        m = re.search(r"Status:\s*(Scored|Field-test)", blob, re.I)
        if m:
            status = "Scored" if m.group(1) == "Scored" else "Field-test"
        m = re.search(r"Task:\s*([A-Z0-9\-]+)", blob, re.I)
        if m:
            task = m.group(1).upper()
        m = re.search(r"Domain:\s*([A-Z]+)", blob, re.I)
        if m:
            domain = m.group(1).upper()
        table_meta.append(
            {
                "correct_option": key,
                "item_status": status,
                "task": task,
                "domain": domain,
            }
        )
    return table_meta, control_bits


def parse_key_items(paras: list[str]) -> tuple[dict[int, dict], dict[int, str]]:
    key_items: dict[int, dict] = {}
    case_keys: dict[int, str] = {}
    current_q = None
    current_case = 0

    def flush() -> None:
        nonlocal current_q
        if current_q:
            key_items[int(current_q["number"])] = current_q
            current_q = None

    for raw in paras:
        line = raw.strip()
        if not line:
            continue
        m = re.match(r"^CASE\s+(\d+)\b", line, re.I)
        if m:
            flush()
            current_case = int(m.group(1))
            continue
        m = re.search(r"Final key:\s*([A-D]{13})", line, re.I)
        if m:
            case_keys[current_case] = m.group(1).upper()
            continue
        m = re.match(r"^Q(\d+)\.\s+(.+)$", line)
        if m:
            flush()
            current_q = {
                "number": int(m.group(1)),
                "case_number": current_case,
                "question_text": m.group(2).strip(),
                "options": {},
                "why_best": "",
                "why_less": {},
                "transfer": "",
            }
            continue
        if current_q:
            m = re.match(r"^([A-D])\.\s+(.+)$", line)
            if m:
                current_q["options"][m.group(1).lower()] = m.group(2).strip()
                continue
            if line.lower().startswith("why the keyed answer is best:"):
                current_q["why_best"] = line[len("Why the keyed answer is best:") :].strip()
                continue
            m = re.match(r"^Why ([A-D]) is less appropriate:\s*(.+)$", line, re.I)
            if m:
                current_q["why_less"][m.group(1).upper()] = m.group(2).strip()
                continue
            if line.lower().startswith("transfer rule:"):
                current_q["transfer"] = line[len("Transfer rule:") :].strip()
                continue
            if re.match(r"^SECTION\s+\d+", line, re.I):
                flush()
                continue
    flush()
    return key_items, case_keys


def main() -> int:
    if not CAND_JSON.is_file() or not KEY_XML.is_file():
        print("Missing candidate JSON or key document.xml", file=sys.stderr)
        return 1

    candidate = json.loads(CAND_JSON.read_text(encoding="utf-8"))
    cand_items = flatten_candidate(candidate)
    paras = extract_paras()
    table_meta, control_bits = parse_table_meta()
    key_items, case_keys = parse_key_items(paras)

    problems: list[str] = []
    if len(cand_items) != 143:
        problems.append(f"Candidate flatten count={len(cand_items)} expected 143")
    if len(table_meta) != 143:
        problems.append(f"Item table count={len(table_meta)} expected 143")
    if len(key_items) != 143:
        problems.append(f"Key parsed questions={len(key_items)} expected 143")

    pass_mentions = [
        re.sub(r"\b[A-D]{11,15}\b", "[KEY-REDACTED]", blob)
        for blob in control_bits
        if re.search(r"pass(ing)?|cut.?score|70\s*%|threshold", blob, re.I)
    ]

    scored_n = field_n = 0
    letter_dist = {k: 0 for k in "abcd"}
    scored_dist = {k: 0 for k in "abcd"}
    domain_dist: dict[str, int] = {}
    admin_rows: dict[str, dict] = {}

    for i in range(143):
        num = i + 1
        cand = cand_items[i]
        key = key_items.get(num)
        meta = table_meta[i]
        code = f"CTA-LPCC-NCMHCE-FB-V2-{num:03d}"

        if not key:
            problems.append(f"Q{num}: missing from answer key parse")
            continue
        if int(cand["number"]) != num:
            problems.append(f"Q{num}: candidate number is {cand['number']}")
        if norm(cand["question_text"]) != norm(key["question_text"]):
            problems.append(f"Q{num}: stem mismatch")
        for opt in "abcd":
            oc = norm(cand["options"].get(opt, ""))
            ok = norm(key["options"].get(opt, ""))
            if not oc or not ok:
                problems.append(f"Q{num}: missing option {opt.upper()}")
            elif oc != ok:
                problems.append(f"Q{num}: option {opt.upper()} mismatch")

        table_letter = str(meta["correct_option"])
        case_no = int(cand["case_number"])
        pos_in_case = (num - 1) % 13
        string_letter = ""
        if case_no in case_keys:
            string_letter = case_keys[case_no][pos_in_case].lower()

        less_letters = list(key["why_less"].keys())
        inferred = [x for x in "ABCD" if x not in less_letters]
        inferred_l = inferred[0].lower() if len(inferred) == 1 else ""

        if table_letter != string_letter and string_letter:
            problems.append(f"Q{num}: table key disagrees with case Final key string")
        if inferred_l and inferred_l != table_letter:
            problems.append(f"Q{num}: inferred letter disagrees with table key")
        if table_letter not in "abcd":
            problems.append(f"Q{num}: invalid table correct letter")

        status = str(meta["item_status"])
        if status == "Scored":
            scored_n += 1
            scored_dist[table_letter] = scored_dist.get(table_letter, 0) + 1
            dom = str(meta["domain"])
            domain_dist[dom] = domain_dist.get(dom, 0) + 1
        elif status == "Field-test":
            field_n += 1
        else:
            problems.append(f"Q{num}: missing Scored/Field-test status")

        letter_dist[table_letter] = letter_dist.get(table_letter, 0) + 1

        if not key["why_best"].strip():
            problems.append(f"Q{num}: missing why_best")
        if len(key["why_less"]) != 3:
            problems.append(f"Q{num}: expected 3 less-appropriate lines, got {len(key['why_less'])}")
        if not key["transfer"].strip():
            problems.append(f"Q{num}: missing transfer rule")

        expl_parts = []
        if key["why_best"].strip():
            expl_parts.append("Why the keyed answer is best: " + key["why_best"].strip())
        for letter in "ABCD":
            if letter in key["why_less"]:
                expl_parts.append(f"Why {letter} is less appropriate: {key['why_less'][letter]}")
        if key["transfer"].strip():
            expl_parts.append("Transfer rule: " + key["transfer"].strip())

        admin_rows[code] = {
            "correct_option": table_letter,
            "item_status": status,
            "domain": str(meta["domain"]),
            "task": str(meta["task"]),
            "source_status": "EXACT/FROZEN SOURCE",
            "explanation": "\n\n".join(expl_parts),
        }

    control_joined = " ".join(control_bits)
    if scored_n != 100 or field_n != 43:
        problems.append(f"Scored/field-test counts {scored_n}/{field_n} (expected 100/43)")
    if scored_dist != {"a": 25, "b": 25, "c": 25, "d": 25}:
        problems.append(f"Scored letter distribution {scored_dist} (expected 25 each)")

    report = [
        "LPCC NCMHCE Form B v2.0 answer-key match report",
        f"candidate_questions={len(cand_items)}",
        f"key_questions={len(key_items)}",
        f"item_tables={len(table_meta)}",
        f"scored={scored_n} field_test={field_n}",
        f"all_letter_dist={json.dumps(letter_dist)}",
        f"scored_letter_dist={json.dumps(scored_dist)}",
        f"scored_domain_dist={json.dumps(domain_dist)}",
        f"passing_percentage_in_source={'KEYWORD_HIT' if pass_mentions else 'NOT_FOUND'}",
        f"control_table_has_100_scored={'yes' if '100 scored-style' in control_joined else 'no'}",
        f"control_table_has_43_field_test={'yes' if '43 field-test-style' in control_joined else 'no'}",
        "rationale_release_in_key_doc=NOT_STATED",
        f"mismatch_count={len(problems)}",
    ]
    for p in problems:
        report.append("MISMATCH: " + p)
    REPORT.write_text("\n".join(report) + "\n", encoding="utf-8")
    print("\n".join(report))

    if problems:
        print("Refusing to write seeds until mismatches are zero.", file=sys.stderr)
        return 1

    header = """<?php
/**
 * ADMIN ONLY — LPCC NCMHCE Form B v2.0 secured answer keys (143 items).
 * Merged into runtime quiz rows by CTA_Lpcc_Ncmhce_Form_B_V2_Answer_Sync only.
 * Never registered as a learner download or exposed via learner AJAX before full submission.
 *
 * @package CTA_LMS
 */
if ( ! defined( 'ABSPATH' ) ) {
\texit;
}

return """
    ADMIN_PATH.parent.mkdir(parents=True, exist_ok=True)
    ADMIN_PATH.write_text(header + php_export(admin_rows) + ";\n", encoding="utf-8")
    print(f"wrote {ADMIN_PATH}")
    print("OK")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/build-lpcc-ncmhce-form-b-v2-seeds.php
<?php
/**
 * Build LPCC NCMHCE Form B v2.0 admin-only answer key from parsed sources.
 *
 * Form B controlled key uses a paragraph format (not Form A item tables).
 *
 * Usage: php scripts/_extract_lpcc_form_b_v2_key_paras.php
 *        php scripts/build-lpcc-ncmhce-form-b-v2-seeds.php
 */

if ( PHP_SAPI !== 'cli' ) {
	fwrite( STDERR, "CLI only.\n" );
	exit( 1 );
}

$root        = dirname( __DIR__ );
$cand_json   = $root . '/_tmp_lpcc_form_b_v2/parsed.json';
$key_paras   = $root . '/_tmp_lpcc_form_b_v2_key/paras.txt';
$key_xml     = $root . '/_tmp_lpcc_form_b_v2_key/unzip/word/document.xml';
$report_path = $root . '/_tmp_lpcc_form_b_v2_key/match_report.txt';

if ( ! is_readable( $cand_json ) || ! is_readable( $key_paras ) || ! is_readable( $key_xml ) ) {
	fwrite( STDERR, "Missing parsed candidate JSON or key extract.\n" );
	exit( 1 );
}

$candidate = json_decode( file_get_contents( $cand_json ), true );
$paras     = file( $key_paras, FILE_IGNORE_NEW_LINES );
if ( ! is_array( $candidate ) || ! is_array( $paras ) ) {
	fwrite( STDERR, "Invalid source files.\n" );
	exit( 1 );
}

if ( ! function_exists( 'fb_v2_norm' ) ) {
	function fb_v2_norm( $text ) {
		$text = html_entity_decode( (string) $text, ENT_QUOTES | ENT_HTML5, 'UTF-8' );
		$map  = array(
			"\xE2\x80\x98" => "'",
			"\xE2\x80\x99" => "'",
			"\xE2\x80\x9C" => '"',
			"\xE2\x80\x9D" => '"',
			"\xC2\xA0"     => ' ',
			'—'            => '-',
			'–'            => '-',
		);
		$text = strtr( $text, $map );
		$text = preg_replace( '/\s+/u', ' ', $text );
		return trim( (string) $text );
	}
}

if ( ! function_exists( 'fb_v2_export' ) ) {
	function fb_v2_export( $value, $level = 0 ) {
		$pad  = str_repeat( "\t", $level );
		$pad1 = str_repeat( "\t", $level + 1 );
		if ( is_array( $value ) ) {
			if ( array() === $value ) {
				return 'array()';
			}
			$is_list = array_keys( $value ) === range( 0, count( $value ) - 1 );
			$out     = "array(\n";
			foreach ( $value as $k => $v ) {
				$key = $is_list ? '' : var_export( (string) $k, true ) . ' => ';
				$out .= $pad1 . $key . fb_v2_export( $v, $level + 1 ) . ",\n";
			}
			return $out . $pad . ')';
		}
		if ( is_bool( $value ) ) {
			return $value ? 'true' : 'false';
		}
		if ( is_int( $value ) || is_float( $value ) ) {
			return (string) $value;
		}
		return var_export( (string) $value, true );
	}
}

$cand_items = array();
foreach ( (array) ( $candidate['cases'] ?? array() ) as $case ) {
	foreach ( (array) ( $case['sections'] ?? array() ) as $sec ) {
		foreach ( (array) ( $sec['questions'] ?? array() ) as $q ) {
			$cand_items[] = array(
				'number'        => (int) ( $q['number'] ?? 0 ),
				'question_text' => (string) ( $q['question_text'] ?? '' ),
				'options'       => (array) ( $q['options'] ?? array() ),
			);
		}
	}
}

if ( 143 !== count( $cand_items ) ) {
	fwrite( STDERR, 'Candidate flatten count=' . count( $cand_items ) . " (expected 143)\n" );
	exit( 1 );
}

$compact_answers = array();
$xml             = file_get_contents( $key_xml );
$xml             = preg_replace( '/w:rsid[A-Za-z]*="[^"]*"/', '', $xml );
$dom             = new DOMDocument();
$dom->loadXML( $xml );
$xpath           = new DOMXPath( $dom );
$xpath->registerNamespace( 'w', 'http://schemas.openxmlformats.org/wordprocessingml/2006/main' );
$cell_text       = static function ( DOMXPath $xpath, DOMNode $cell ) {
	$texts = $xpath->query( './/w:t', $cell );
	$line  = '';
	foreach ( $texts as $t ) {
		$line .= $t->textContent;
	}
	return trim( preg_replace( '/\s+/u', ' ', $line ) );
};
$tables = $xpath->query( '//w:tbl' );
foreach ( $tables as $tbl ) {
	$rows = $xpath->query( './w:tr', $tbl );
	foreach ( $rows as $tr ) {
		$cells = $xpath->query( './w:tc', $tr );
		$vals  = array();
		foreach ( $cells as $tc ) {
			$vals[] = $cell_text( $xpath, $tc );
		}
		for ( $i = 0; $i + 1 < count( $vals ); $i += 2 ) {
			if ( preg_match( '/^\d+$/', (string) $vals[ $i ] ) && preg_match( '/^[A-D]$/i', (string) $vals[ $i + 1 ] ) ) {
				$compact_answers[ (int) $vals[ $i ] ] = strtolower( $vals[ $i + 1 ] );
			}
		}
	}
}

$key_items    = array();
$current      = null;
$current_case = 0;
$flush        = static function () use ( &$key_items, &$current ) {
	if ( $current ) {
		$key_items[ (int) $current['number'] ] = $current;
		$current = null;
	}
};

foreach ( $paras as $raw ) {
	$line = trim( (string) $raw );
	if ( '' === $line ) {
		continue;
	}
	if ( preg_match( '/^Case\s+(\d+):/i', $line, $m ) ) {
		$flush();
		$current_case = (int) $m[1];
		continue;
	}
	if ( preg_match( '/^(\d{1,3})\.\s+(.+)$/u', $line, $m ) && (int) $m[1] <= 143 ) {
		$flush();
		$current = array(
			'number'        => (int) $m[1],
			'case_number'   => $current_case,
			'question_text' => trim( $m[2] ),
			'correct_option'=> '',
			'item_status'   => '',
			'domain'        => '',
			'task'          => '',
			'why_best'      => '',
			'why_not'       => array(),
			'transfer'      => '',
		);
		continue;
	}
	if ( ! $current ) {
		continue;
	}
	if ( preg_match( '/^Correct Answer:\s*([A-D])\./i', $line, $m ) ) {
		$current['correct_option'] = strtolower( $m[1] );
		continue;
	}
	if ( preg_match( '/^Domain:\s*([A-Z]+)/i', $line, $m ) ) {
		$current['domain'] = strtoupper( $m[1] );
	}
	if ( preg_match( '/Task:\s*([A-Z0-9\-]+)/i', $line, $m ) ) {
		$current['task'] = strtoupper( $m[1] );
	}
	if ( preg_match( '/Blueprint role:\s*(scored-style|field-test-style)/i', $line, $m ) ) {
		$current['item_status'] = ( false !== stripos( $m[1], 'field' ) ) ? 'Field-test' : 'Scored';
	}
	if ( 0 === stripos( $line, 'Why it is best:' ) ) {
		$current['why_best'] = trim( substr( $line, strlen( 'Why it is best:' ) ) );
		continue;
	}
	if ( preg_match( '/^Why ([A-D]) is not best:\s*(.+)$/u', $line, $m ) ) {
		$current['why_not'][ strtoupper( $m[1] ) ] = trim( $m[2] );
		continue;
	}
	if ( 0 === stripos( $line, 'Clinical discrimination / transfer rule:' ) ) {
		$current['transfer'] = trim( substr( $line, strlen( 'Clinical discrimination / transfer rule:' ) ) );
		continue;
	}
}
$flush();

$problems    = array();
$scored_n    = 0;
$field_n     = 0;
$scored_dist = array( 'a' => 0, 'b' => 0, 'c' => 0, 'd' => 0 );
$domain_dist = array();
$admin_rows  = array();
$pass_mentions = array();
foreach ( $paras as $raw ) {
	if ( preg_match( '/pass(ing)?|cut.?score|70\s*%|threshold/i', (string) $raw ) ) {
		$pass_mentions[] = (string) $raw;
	}
}

if ( 143 !== count( $key_items ) ) {
	$problems[] = 'Key parsed questions=' . count( $key_items ) . ' expected 143';
}
if ( 143 !== count( $compact_answers ) ) {
	$problems[] = 'Compact answer grid count=' . count( $compact_answers ) . ' expected 143';
}

for ( $i = 0; $i < 143; $i++ ) {
	$num  = $i + 1;
	$cand = $cand_items[ $i ];
	$key  = isset( $key_items[ $num ] ) ? $key_items[ $num ] : null;
	$code = sprintf( 'CTA-LPCC-NCMHCE-FB-V2-%03d', $num );

	if ( ! $key ) {
		$problems[] = "Q{$num}: missing from answer key parse";
		continue;
	}
	if ( (int) $cand['number'] !== $num ) {
		$problems[] = "Q{$num}: candidate number is {$cand['number']}";
	}
	if ( fb_v2_norm( $cand['question_text'] ) !== fb_v2_norm( $key['question_text'] ) ) {
		$problems[] = "Q{$num}: stem mismatch";
	}

	$letter = (string) $key['correct_option'];
	if ( ! in_array( $letter, array( 'a', 'b', 'c', 'd' ), true ) ) {
		$problems[] = "Q{$num}: invalid correct letter";
	}
	if ( isset( $compact_answers[ $num ] ) && $compact_answers[ $num ] !== $letter ) {
		$problems[] = "Q{$num}: compact grid letter disagrees with Correct Answer line";
	}

	$status = (string) $key['item_status'];
	if ( 'Scored' === $status ) {
		++$scored_n;
		++$scored_dist[ $letter ];
		$dom = (string) $key['domain'];
		if ( ! isset( $domain_dist[ $dom ] ) ) {
			$domain_dist[ $dom ] = 0;
		}
		++$domain_dist[ $dom ];
	} elseif ( 'Field-test' === $status ) {
		++$field_n;
	} else {
		$problems[] = "Q{$num}: missing scored/field-test blueprint role";
	}

	if ( '' === trim( (string) $key['why_best'] ) ) {
		$problems[] = "Q{$num}: missing why_best";
	}
	if ( 3 !== count( $key['why_not'] ) ) {
		$problems[] = "Q{$num}: expected 3 why-not lines, got " . count( $key['why_not'] );
	}
	if ( '' === trim( (string) $key['transfer'] ) ) {
		$problems[] = "Q{$num}: missing transfer rule";
	}

	$expl_parts = array();
	if ( '' !== trim( (string) $key['why_best'] ) ) {
		$expl_parts[] = 'Why the keyed answer is best: ' . trim( $key['why_best'] );
	}
	foreach ( array( 'A', 'B', 'C', 'D' ) as $opt_letter ) {
		if ( isset( $key['why_not'][ $opt_letter ] ) ) {
			$expl_parts[] = 'Why ' . $opt_letter . ' is less appropriate: ' . $key['why_not'][ $opt_letter ];
		}
	}
	if ( '' !== trim( (string) $key['transfer'] ) ) {
		$expl_parts[] = 'Transfer rule: ' . trim( $key['transfer'] );
	}

	$admin_rows[ $code ] = array(
		'correct_option' => $letter,
		'item_status'    => $status,
		'domain'         => (string) $key['domain'],
		'task'           => (string) $key['task'],
		'source_status'  => 'EXACT/FROZEN SOURCE',
		'explanation'    => implode( "\n\n", $expl_parts ),
	);
}

if ( 100 !== $scored_n || 43 !== $field_n ) {
	$problems[] = "Scored/field-test counts {$scored_n}/{$field_n} (expected 100/43)";
}

$report   = array();
$report[] = 'LPCC NCMHCE Form B v2.0 answer-key match report';
$report[] = 'candidate_questions=' . count( $cand_items );
$report[] = 'key_questions=' . count( $key_items );
$report[] = 'compact_grid=' . count( $compact_answers );
$report[] = 'scored=' . $scored_n . ' field_test=' . $field_n;
$report[] = 'scored_letter_dist=' . json_encode( $scored_dist );
$report[] = 'scored_domain_dist=' . json_encode( $domain_dist );
$report[] = 'passing_percentage_in_source=' . ( $pass_mentions ? 'KEYWORD_HIT' : 'NOT_FOUND' );
$report[] = 'rationale_release_in_key_doc=AFTER_FULL_FORM_COMPLETION_ONLY';
$report[] = 'mismatch_count=' . count( $problems );
foreach ( $problems as $p ) {
	$report[] = 'MISMATCH: ' . $p;
}
file_put_contents( $report_path, implode( "\n", $report ) . "\n" );
echo implode( "\n", $report ) . "\n";

if ( $problems ) {
	fwrite( STDERR, "Refusing to write seeds until mismatches are zero.\n" );
	exit( 1 );
}

$admin_path = $root . '/includes/quiz-seeds/admin-only/lpcc-ncmhce-form-b-v2-answer-key.php';
$admin_php  = "<?php\n";
$admin_php .= "/**\n";
$admin_php .= " * ADMIN ONLY — LPCC NCMHCE Form B v2.0 secured answer keys (143 items).\n";
$admin_php .= " * Merged into runtime quiz rows by CTA_Lpcc_Ncmhce_Form_B_V2_Answer_Sync only.\n";
$admin_php .= " * Never registered as a learner download or exposed via learner AJAX before full submission.\n";
$admin_php .= " *\n * @package CTA_LMS\n */\n";
$admin_php .= "if ( ! defined( 'ABSPATH' ) ) {\n\texit;\n}\n\n";
$admin_php .= 'return ' . fb_v2_export( $admin_rows ) . ";\n";
file_put_contents( $admin_path, $admin_php );
echo "wrote {$admin_path}\n";
echo "OK\n";
